include/xtensor-signal/find_peaks.hpp

```diff
@@ -0,0 +1,725 @@
+            {
+                //if the value is a none type
+                if(!value.has_value())
+                {
+                    return -1;
+                }
+                else
+                {
+                    //otherwise we have a number
+                    //could probably add a check for arithmatic type here
+                    if (1 < value.value())
+                    {
+                        return std::ceil(value.value());
+                    }
+                }
+
+                throw std::runtime_error("wlen must be greater thank 1");
+
+                return 0;
+            }
+
+            /**
+             * @brief Calculate the width of each peak in a signal.
+             * @details This function calculates the width of a peak in samples at a relative
+             *     distance to the peak's height and prominence.
+             * @param x A signal with peaks.
+             * @param peaks Indices of peaks in `x`.
+             * @param rel_height
+             *     Chooses the relative height at which the peak width is measured as a
+             *     percentage of its prominence. 1.0 calculates the width of the peak at
+             *     its lowest contour line while 0.5 evaluates at half the prominence
+             *     height. Must be at least 0. See notes for further explanation.
+             * @param prominence_data 
+             *     A tuple of three arrays matching the output of `peak_prominences` when
+             *     called with the same arguments `x` and `peaks`. This data are calculated
+             *     internally if not provided.
+             * @param wlen 
+             *     A window length in samples passed to `peak_prominences` as an optional
+             *     argument for internal calculation of `prominence_data`. This argument
+             *     is ignored if `prominence_data` is given.
+             * @returns widths The widths for each peak in samples. width_heights The height of the contour lines at which the `widths` where evaluated. left_ips, right_ips 
+                    Interpolated positions of left and right intersection points of a horizontal line at the respective evaluation height.
+             * @note Derived from https://github.com/scipy/scipy/blob/main/scipy/signal/_peak_finding_utils.pyx
+             */
+            template <
+                class E1,
+                class E2,
+                class E3,
+                class E4,
+                class E5,
+                class E6>
+                auto peak_widths(
+                    E1&& x,
+                    E2&& peaks,
+                    E3&& rel_height,
+                    E4&& prominences,
+                    E5&& left_bases,
+                    E6&& right_bases)
+            {
+                if (rel_height < 0)
+                {
+                    throw std::runtime_error(
+                        "Relative height must be equal to or greater than 0");
+                }
```

Comment:
I think you cna use `XGTENSOR_ASSERT` here. This will make the code more concise and will allow to disable the checks.